[Bigtable] Documented defaults timeout values in the EnhancedBigtableStubSettings

rahulKQL opened the pull request:
Fixes #2992

Updated default timeouts for all the operations in `EnhnaceDBigtableStubSettings` as `BigtableClientSettings` settings were deprecated.

rahulKQL commented:
@igorbernstein2 Please have a look at this change. 
Please let me know if the format can be improved for timeout listing.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -55,6 +55,24 @@
+ *         #mutateRowSettings()}, {@link #bulkMutateRowsSettings()} {@link
+ *         #readModifyWriteRowSettings()}:
+ *         <li>Default timeout for {@link RetrySettings#getInitialRpcTimeout()} is 20_000ms.
+ *         <li>Default timeout for {@link RetrySettings#getMaxAttempts()} is 20_000ms.
```
timeout for max attempts as a duration?

rahulKQL replied:
My apologies, It was a typo from my end it should be `getMaxRpcTimeout`.

igorbernstein2 replied:
I would collapse the rpc timeout min/max/multiplier settings into a single constant rpc timeout value in the docs

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -55,6 +55,24 @@
  *   <li>The default service address (bigtable.googleapis.com) and default port (443) are used.
  *   <li>Credentials are acquired automatically through Application Default Credentials.
  *   <li>Retries are configured for idempotent methods but not for non-idempotent methods.
+ *   <li>These are the default timeouts used Bigtable operations.
```
This needs a lot of work and should probably be in its own paragraph and/or in each method doc:
- A distinction should be drawn between idempotent methods (ReadRow(s), MutateRow(s), SampleRowKeys) and non-idempotent methods (CheckAndMutateRows, ReadModiifyWrite)
- The distinction for streaming vs non-streaming should be kept. But there needs to be some explanation about how the settings apply. For example rpc timeout for non-streaming is the deadline for each call, but for streaming the rpc timeout is the timeout between each row. 
- There should be docs explaining what rpc timeout is vs total timeout (ie. attempt timeout vs operation timeout)
- The values should be human readable, ie instead of 20_000ms, it should be 20 seconds
- There should be docs about exponential delay defaults 
- There should be an explanation of what idleTimeout is 
- The formatting is currently invalid: ```<ul>TEXT<li>first element text``` is not valid markup

igorbernstein commented:
@kennethlowery8190, please remove the approval. This pr needs more work and I don’t want it accidentally merged

rahulKQL commented:
@igorbernstein2  Please have another round of review. 

I have added below points in the last commit.
- Listed Description for streaming and non-streaming operation.
- Tried explaining RPC vs Total timeout
- Tried explaining retryDelayMultiplier.
- Added default values on each setting operation's Javadoc.
- Added bulkMutateRowsSettings description.

sduskis commented:
@igorbernstein2 or @kolea2, can you PTAL

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -55,6 +55,13 @@
+ *   <li>Each operation setting accepts RPC and total time out. Here, RPC timeouts are for each
+ *       attempt to get the result whereas total timeout applies to the operation timeout.
+ *   <li>RetryDelayMultiplier controls the change in retry delay. After initial attempt, subsequent
+ *       attempts are calculated by multiplying the retry delay of the previous call.
```
This feels out of place. The first  3 bullet points describe the values of the settings, while new 3 points are describing concepts. Also, what are you referring to by `this settings`? Also, it's weird that the explanation for RetryDelayMultiplier is 3 bullet points away from the Retries bullet point.

I think it would be better to link users to read the docs for RetrySettings for a description of the concepts and focus on the values in these docs. (see below)

rahulKQL replied:
I think with your new suggested explanation, User will get the idea about these Retries and RPC timeouts so I have removed these explanations from the Class level Javadoc. 
Please let me know if we want to add any details here.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   *
+   * <p>The idle timeout is configured via {@code ServerStreamingCallSettings#getIdleTimeout}. This
+   * timer will terminate any stream that has not has seen any demand or activity in the configured
+   * interval. To turn off idle checks, set the idleTimeout to {@link Duration#ZERO}.
```
Please document the default value here (5 minutes)

rahulKQL replied:
I have rephrased this line with default time.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   * Code#UNAVAILABLE} failure code.
+   *
+   * <p>RetryDelay starts with 100ms or 0.1 seconds, which multiplies with 1.3 times on every
+   * reattempt till a max retry timeout value of 60 seconds.
```
Please drop `0.1 seconds`. Also, this isnt a timeout. Maybe reword to:

The RetryDelay between failed attempts starts with 100ms and exponentially increases by a factor of 1.3 until 60 seconds.

rahulKQL replied:
Thanks for the detailed suggestion, I have updated these settings with the same.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   * <p>RetryDelay starts with 100ms or 0.1 seconds, which multiplies with 1.3 times on every
+   * reattempt till a max retry timeout value of 60 seconds.
+   *
+   * <p>RPC timeout is 20 seconds.
```
I think you should combine the 2 timeouts docs and call out that the RPC timeout means wait timeout:

The default read timeout for each row in a response stream is 20 seconds. And the timeout to read the entire stream is 1 hour.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   *
+   * <p>RPC timeout is 20 seconds.
+   *
+   * <p>Total timeout for this operation is 600 seconds or 10 mins.
```
Similar to the streaming example I think you should combine the timeouts:

The default timeout for  each attempt is 20 seconds and the timeout for the entire operation across all of the attempts is 10 mins.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   *
+   * <p>This is idempotent and streaming operation.
+   *
+   * <p>The idle timeout is configured via {@code ServerStreamingCallSettings#getIdleTimeout}. This
```
Maybe nest these in a list? Also as opposed to duplicating the docs in RetrySettings and ServerStreamingSettings, just link out to them. Also I think it would better to link directly to the setters

Maybe something this?

```
<p>Default settings:
<ul>
  <li>{@link ServerStreamingCallSettings.Builder#setIdleTimeout Default idle timeout} is set to 1 hour
  <li>{@link ServerStreamingCallSettings.Builder#setRetryableCodes Error codes} are: {@link Code#DEADLINE_EXCEEDED} and {@link Code#UNAVAILABLE}
<li>RetryDelay {@link RetrySettings.Builder#setInitialRetryDelay starts} at 100ms and {@link RetrySettings.Builder#setRetryDelayMultiplier increases exponentially} by 1.3 until a maximum of ....


```

rahulKQL commented:
@igorbernstein2 I tried to address feedback comments, Please take a look.

igorbernstein2 commented:
Just a couple minor nits, but LGTM!

rahulKQL commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +185,100 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   *       5 mins.
+   *   <li>Retry {@link ServerStreamingCallSettings.Builder#setRetryableCodes error codes} are:
+   *       {@link Code#DEADLINE_EXCEEDED} and {@link Code#UNAVAILABLE}.
+   *   <li>RetryDelay between failed attempt {@link RetrySettings.Builder#setInitialRetryDelay
```
Thanks for the review, Have fixed this typo.

rahulKQL commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -185,22 +185,100 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
+   *       exponentially} by a factor of 1.3 until a {@link RetrySettings.Builder#setMaxRetryDelay
+   *       maximum of} 60 seconds.
+   *   <li>The default timeout for {@link RetrySettings.Builder#setMaxRpcTimeout each attempt} is 20
+   *       seconds and the timeout for {@link RetrySettings.Builder#setTotalTimeout entire
```
Addressed this as well.

rahulKQL commented:
@igorbernstein2 Have addressed your feedback, Please have another look.

igorbernstein2 commented:
Look like a spanner test is failing
